lib/vnet/diag/routeconflict.go

```diff
@@ -0,0 +1,302 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package diag
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"net"
+	"net/netip"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+)
+
+var log = logutils.NewPackageLogger(teleport.ComponentKey, "vnet:diag")
+
+type RouteConflictConfig struct {
+	VnetIfaceName string
+	Routing       Routing
+	Interfaces    Interfaces
+}
+
+// Routing abstracts away platform-specific logic of obtaining routes with their destinations,
+// allowing running tests for the general logic behind RouteConflictDiag on any platform.
+type Routing interface {
+	GetRouteDestinations() ([]RouteDest, error)
+}
+
+// Interfaces abstracts away functions from the net package and calls to ifconfig, allowing mocking
+// interactions with them in tests.
+type Interfaces interface {
+	InterfaceByName(string) (*net.Interface, error)
+	InterfaceByIndex(int) (*net.Interface, error)
+	// InterfaceApp attempts to return the name of the app that created the interface given the name
+	// of the interface.
+	InterfaceApp(context.Context, string) (string, error)
+}
+
+func (c *RouteConflictConfig) CheckAndSetDefaults() error {
+	if c.VnetIfaceName == "" {
+		return trace.BadParameter("missing VNet interface name")
+	}
+
+	if c.Routing == nil {
+		return trace.BadParameter("missing routing")
+	}
+
+	if c.Interfaces == nil {
+		return trace.BadParameter("missing net interfaces")
+	}
+
+	return nil
+}
+
+type RouteConflictDiag struct {
+	cfg *RouteConflictConfig
+}
+
+func NewRouteConflictDiag(cfg *RouteConflictConfig) (*RouteConflictDiag, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &RouteConflictDiag{
+		cfg: cfg,
+	}, nil
+}
+
+// Run scans Ipv4 routing table (equivalent of running `netstat -rn -f inet`) in search of routes
+// that overlap with routes set up by VNet.
+//
+// If a 3rd-party route conflicts with more than one VNet route, Run returns a single RouteConflict
+// for that 3rd-party route describing the conflict with the first conflicting VNet route.
+func (c *RouteConflictDiag) Run(ctx context.Context) ([]RouteConflict, error) {
+	retries := 0
+	for {
+		crs, err := c.run(ctx)
+		if err != nil {
+			// UnstableIfaceError usually means that an interface was removed between fetching route
+			// messages and getting the details of the interface. In this case, the routes for that
+			// interface are likely gone too, so the best course of action is to repeat the whole check.
+			if errors.As(err, new(UnstableIfaceError)) && retries < 2 {
```

Comment:
I saw people do this when I was browsing other code that uses `golang.org/x/net/route`.